### PR TITLE
Announcing widget errors and warnings

### PR DESCRIPTION
**Pull request by guimafelipe**
## Summary of the pull request
When we have a warning inside a widget, or an error message, the narrator should announce this event to the user, reading the message for them. 
## References and relevant issues
https://dev.azure.com/microsoft/OS/_workitems/edit/48793024
## Detailed description of the pull request / Additional comments
This changes adds a linear search on the widget whenever it is rendered, looking for potential warning messages, which here are any TextBlocks inside containers with `warning` style. When found, we use the announcement service to narrate it to the user.

In the case of an error card, all the message is announced to the user when rendered.
## PR checklist
- [ ] Closes #xxx
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

File: tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs [hunk 1]
```diff
@@ -339,4 +348,57 @@ public void UnsubscribeFromWidgetUpdates()
...
+        {
+            SearchForWarning(element, false);
+        }
+    }
+
+    private void SearchForWarning(IAdaptiveCardElement element, bool isInsideWarningContainer)
+    {
+        // We are only interested in plain texts. Buttons, Actions, Images
+        // and textboxes are all ignored. Including ActionSets and ImageSets.
+        if (element is AdaptiveTextBlock)
```
Comment (dhoehna) [inline tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:364]: The new variable name can be added on this line.  `element is AdaptiveTextBlock textBlock`

Comment (dhoehna) [reply tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:364]: Same for all the other if's

File: tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs [hunk 2]
```diff
@@ -339,4 +348,57 @@ public void UnsubscribeFromWidgetUpdates()
...
+
+    private void AnnounceWarnings(AdaptiveCard card)
+    {
+        foreach (var element in card.Body)
+        {
+            SearchForWarning(element, false);
+        }
+    }
+
+    private void SearchForWarning(IAdaptiveCardElement element, bool isInsideWarningContainer)
```
Comment (dhoehna) [inline tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:360]: This method recursively searches through every element in a widget to find every container with a "warning" style.  Does that sound right?

File: tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs [hunk 3]
```diff
@@ -339,4 +348,57 @@ public void UnsubscribeFromWidgetUpdates()
...
+    {
+        // We are only interested in plain texts. Buttons, Actions, Images
+        // and textboxes are all ignored. Including ActionSets and ImageSets.
+        if (element is AdaptiveTextBlock)
+        {
+            var textBlock = (AdaptiveTextBlock)element;
+            if (isInsideWarningContainer)
+            {
+                _screenReaderService.Announce(textBlock.Text);
```
[the remark this answers is not in the document]
Comment (guimafelipe) [reply tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:368]: Done

File: tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs (hunk 2 above)
Comment (guimafelipe) [reply tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:360]: Yes, this assumes that any text inside a warning style container should be announced on a widget.

File: tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs [hunk 4]
```diff
@@ -183,6 +190,7 @@ private async Task RenderWidgetFrameworkElementAsync()
                     {
                         _renderedCard.Action += HandleAdaptiveAction;
                         WidgetFrameworkElement = _renderedCard.FrameworkElement;
+                        AnnounceWarnings(card.AdaptiveCard);
```
Comment (krschau) [inline tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:193]: Can we scope this to only when a screen reader is running? So we don't have to do a ton of tree traversals unnecessarily.
https://www.hanselman.com/blog/programmatically-detecting-screen-readers

File: tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs [hunk 5]
```diff
@@ -339,4 +352,54 @@ public void UnsubscribeFromWidgetUpdates()
...
+            {
+                _screenReaderService.Announce(textBlock.Text);
+            }
+
+            return;
+        }
+
+        if (element is IAdaptiveContainerBase containerElement)
+        {
+            foreach (var containerType in containerTypes)
```
Comment (dhoehna) [inline tools/Dashboard/DevHome.Dashboard/ViewModels/WidgetViewModel.cs:391]: Can be simplified to `if containerTypes.Where(x => x.GetType() == containerType.Key`

**Review by EricJohnson327 — APPROVED** [review]

**Review by dkbennett — APPROVED** [review]

**Pull request merged**